### PR TITLE
Raw prime wf for Run 3 data w/ dropped strips + approx. clusters

### PR DESCRIPTION
**Pull request by mandrenguyen**
#### PR description:

A new RelVal workflow (140.6) is introduced, with the reconstruction sequence intended for the HI commissioning run.   As a placeholder recent pp data is used.  A new `rawPrimeDataRepacker` sequence is introduced to drop the strip info and rename the raw data collection, emulating what will happen at the HLT.  Previously, the approximated clusters were produced, but the strip info was not actually dropped, such that this new raw data label wasn't needed/implemented.

The strip info is also now dropped in the Run 2 raw prime workflow (140.58).   

Finally, the approximated cluster collection is renamed to hltSiStripClusters2ApproxClusters, as this will be the name of the collection coming out of the HLT, as discussed in [JIRA](https://its.cern.ch/jira/browse/CMSHLT-2521)

@icali @abaty @FHead @Martin-Grunewald 

#### PR validation:

Tested w/ wfs:  
140.58, 140.6, 159.02

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

A backport to 12_5_X will be made.

### CODE REVIEW

**Comment by mandrenguyen** [comment]
test parameters:
- workflows=159,159.02,159.03,140.56,140.58,140.6

**Comment by mandrenguyen** [comment]
please test

**Comment by mandrenguyen** [comment]
please test

File: Configuration/StandardSequences/python/DigiToRaw_Repack_cff.py [hunk 1]
```diff
@@ -72,7 +72,14 @@
...
...6, 857, 858, 859, 861, 862, 863, 864, 865, 866, 867, 868, 869, 1024, 1025, 1026, 1027, 1028, 1029, 1030, 1031, 1032, 1033, 1034, 1035, 1036, 1037, 1038, 1039, 1040, 1041, 1042, 1043, 1044, 1045, 1046, 1047, 1048, 1049, 1050, 1051, 1100, 1101, 1102, 1103, 1104, 1105, 1106, 1107, 1108, 1109, 1110, 1111, 1112, 1113, 1114, 1115, 1116, 1117, 1118, 1119, 1120, 1121, 1122, 1123, 1134, 1135, 1200, 1201, 1202, 1203, 1204, 1205, 1206, 1207, 1208, 1209, 1212, 1213, 1214, 1215, 1216, 1217, 1218, 1219, 1220, 1221, 1224, 1225, 1226, 1227, 1228, 1229, 1230, 1231, 1232, 1233, 1236, 1237, 1238, 1239, 1240, 1241, 1242, 1243, 1244, 1245, 1248, 1249, 1250, 1251, 1252, 1253, 1254, 1255, 1256, 1257, 1260, 1261, 1262, 1263, 1264, 1265, 1266, 1267, 1268, 1269, 1272, 1273, 1274, 1275, 1276, 1277, 1278, 1279, 1280, 1281, 1284, 1285, 1286, 1287, 1288, 1289, 1290, 1291, 1292, 1293, 1296, 1297, 1298, 1299, 1300, 1301, 1302, 1308, 1309, 1310, 1311, 1312, 1313, 1314, 1320, 1321, 1322, 1323, 1324, 1325, 1326, 1332, 1333, 1334, 1335, 1336, 1337, 1338, 1354, 1356, 1358, 1360, 1368, 1369, 1370, 1371, 1376, 1377, 1380, 1381, 1384, 1385, 1386, 1390, 1391, 1392, 1393, 1394, 1395, 1402, 1404, 1405, 1462, 1463, 1467 ] ),
```
Comment (mmusich) [inline Configuration/StandardSequences/python/DigiToRaw_Repack_cff.py:80]: would splitting this line make it more readable?
Also for my education, if I understand the logic correctly, this is every other FED in CMS, *excepted* the SiStrip Tracker ones. Why are there holes in the list, e.g. 526,527, etc? Are these currently non-functional FEDs? Wouldn't be more practical to generate the list of numbers in an automatic way, so that is a FED becomes functional in the future this list hasn't to be updated? Thanks for the clarification.

Comment (missirol) [reply Configuration/StandardSequences/python/DigiToRaw_Repack_cff.py:80]: Adding a question for my own education (since this list also gets into the HLT menu, and there it cannot be created dinamically at python level): if the aim is to exclude known FEDs, couldn't we write a simple plugin to do just that (if it doesn't exist already)? e.g.
```python
rawPrimeDataRepacker = cms.EDProducer( "EvFFEDAntiSelector",
  excludedFedList = cms.vuint32( [ .. ] ),
  inputTag = cms.InputTag( "rawDataCollector" )
)
```

Comment (mandrenguyen) [reply Configuration/StandardSequences/python/DigiToRaw_Repack_cff.py:80]: Thanks, I have to defer to @abaty @icali on these points.

File: RecoLocalTracker/SiStripClusterizer/python/SiStripClusters2ApproxClusters_cff.py [hunk 2]
```diff
@@ -1,5 +1,5 @@
 from RecoLocalTracker.SiStripClusterizer.SiStripClusters2ApproxClusters_cfi import *
 
 from Configuration.ProcessModifiers.approxSiStripClusters_cff import approxSiStripClusters
-SiStripClusters2ApproxClustersHLT = SiStripClusters2ApproxClusters.clone()
-approxSiStripClusters.toModify(SiStripClusters2ApproxClustersHLT, inputClusters = "siStripClustersHLT")
+hltSiStripClusters2ApproxClusters = SiStripClusters2ApproxClusters.clone()
+approxSiStripClusters.toModify(hltSiStripClusters2ApproxClusters, inputClusters = "siStripClustersHLT")
```
Comment (missirol) [inline RecoLocalTracker/SiStripClusterizer/python/SiStripClusters2ApproxClusters_cff.py:5]: Why is this needed?

The convention (the way I know it) is that modules inside the HLT menu are normally called `hlt*`, while the opposite applies to modules in other steps of the chain, like this one. (cc: @Martin-Grunewald)

File: Configuration/EventContent/python/EventContent_cff.py [hunk 3]
```diff
@@ -779,7 +787,9 @@ def SwapKeepAndDrop(l):
 REPACKRAWEventContent.outputCommands.extend(HLTriggerRAW.outputCommands)
 approxSiStripClusters.toModify(REPACKRAWEventContent,
                                outputCommands = REPACKRAWEventContent.outputCommands+[
-                                   'keep *_SiStripClusters2ApproxClustersHLT_*_*'
+                                   'keep *_hltSiStripClusters2ApproxClusters_*_*',
+                                   'drop FEDRawDataCollection_rawDataRepacker_*_*',
+                                   'keep *_rawPrimeDataRepacker_*_*'
```
Comment (missirol) [inline Configuration/EventContent/python/EventContent_cff.py:792]: Would this also work?
```suggestion
                                   'keep FEDRawDataCollection_rawPrimeDataRepacker_*_*'
```

File: Configuration/StandardSequences/python/DigiToRaw_Repack_cff.py (hunk 1 above)
Comment (missirol) [reply Configuration/StandardSequences/python/DigiToRaw_Repack_cff.py:80]: I can use a clarification.

> as it was the suggested strategy

Suggested by whom?

> easily accessible also if modulo mistakes

What does this mean?

> the list will never change

And this is guaranteed by whom?

Borrowing questions from upstairs:

> would splitting this line make it more readable?

> Why are there holes in the list, e.g. 526,527, etc? Are these currently non-functional FEDs?

File: RecoLocalTracker/SiStripClusterizer/python/SiStripClusters2ApproxClusters_cff.py (hunk 2 above)
Comment (mandrenguyen) [reply RecoLocalTracker/SiStripClusterizer/python/SiStripClusters2ApproxClusters_cff.py:5]: This is the name of the collection that will be produced at the HLT.  You can see the JIRA ticket linked in the PR description for details.  In this particular workflow we are just creating the collection offline for testing purposes.

File: Configuration/EventContent/python/EventContent_cff.py (hunk 3 above)
Comment (mandrenguyen) [reply Configuration/EventContent/python/EventContent_cff.py:792]: good point, we should specify the type.  Will wait a bit for more comments and commit that, also for  `'keep *_hltSiStripClusters2ApproxClusters_*_*',`

File: Configuration/EventContent/python/EventContent_cff.py [hunk 4]
```diff
@@ -634,6 +638,10 @@ def SwapKeepAndDrop(l):
 FEVTDEBUGHLTEventContent.outputCommands.append('keep *_*_MergedTrackTruth_*')
 FEVTDEBUGHLTEventContent.outputCommands.append('keep *_*_StripDigiSimLink_*')
 FEVTDEBUGHLTEventContent.outputCommands.append('keep *_*_PixelDigiSimLink_*')
+approxSiStripClusters.toModify(FEVTDEBUGHLTEventContent,
+                              outputCommands = FEVTDEBUGHLTEventContent.outputCommands+[
+                                  'keep *_hltSiStripClusters2ApproxClusters_*_*'
```
Comment (missirol) [inline Configuration/EventContent/python/EventContent_cff.py:643]: This seems unnecessary, because `FEVTDEBUGHLTEventContent` is a superset of `FEVTDEBUGEventContent`, and the latter already includes `hltSiStripClusters2ApproxClusters` due to L625.

**Comment by francescobrivio** [comment]
Hi @mandrenguyen!
After a little chat with @mmusich, I think it would be very useful to define a new Era (`Run3_pp_on_PbPb_rawprime`?) which is exactly as `Run3_pp_on_PbPb` but with the addition of the `approxSiStripClusters` modifier.
This (imho) would have two main benefits:
 - DQM clients that make use of tracking could be configured with such era and tracking would be restored even in HI conditions (i.e. when only approxClusters are available)
 - We could then define a new scenario to be used in Tier0 for running the reconstruction of the rawprime data (and not only the repacking)
 
 What do you all think?

**Comment by mandrenguyen** [comment]
> What do you all think?
Hmmmm, good point, i don't recall why I just didn't make a new ere to begin with.  I'll update the PR to include that shortly.

**Comment by mandrenguyen** [comment]
please test

File: Configuration/Eras/python/Era_Run3_pp_on_PbPb_approxSiStripClusters_cff.py [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import FWCore.ParameterSet.Config as cms
+
+from Configuration.Eras.Era_Run3_pp_on_PbPb_cff import Run3_pp_on_PbPb
+from Configuration.Eras.Modifier_tracker_apv_vfp30_2016_cff import tracker_apv_vfp30_2016
```
Comment (mmusich) [inline Configuration/Eras/python/Era_Run3_pp_on_PbPb_approxSiStripClusters_cff.py:4]: Why is this needed?

Comment (mandrenguyen) [reply Configuration/Eras/python/Era_Run3_pp_on_PbPb_approxSiStripClusters_cff.py:4]: oh it's not.  Thanks for spotting it.

**Comment by mandrenguyen** [comment]
please abort

**Comment by mandrenguyen** [comment]
please test

**Comment by mandrenguyen** [comment]
urgent
backport of this PR needed for a new release to be part of the heavy-ion test next Monday (november 7)

File: Configuration/StandardSequences/python/DigiToRaw_Repack_cff.py (hunk 1 above)
Comment (mandrenguyen) [reply Configuration/StandardSequences/python/DigiToRaw_Repack_cff.py:80]: I don't know if @icali and/or @abaty can respond better, but this is how it's done at the HLT.  We are simply reproducing this behavior for testing purposes.  In actual data processing, this code will not be used offline.  So in the interest of moving forward with the HI test, I would propose we go forward with this and reconsider code quality / readability issues for the actual 2023 data taking.

File: RecoLocalTracker/SiStripClusterizer/python/SiStripClusters2ApproxClusters_cff.py (hunk 2 above)
Comment (mandrenguyen) [reply RecoLocalTracker/SiStripClusterizer/python/SiStripClusters2ApproxClusters_cff.py:5]: @missirol @FHead Perhaps you can advise whether we stick with this naming convention or whether the name will change on the HLT side.

Comment (Martin-Grunewald) [reply RecoLocalTracker/SiStripClusterizer/python/SiStripClusters2ApproxClusters_cff.py:5]: Module labels at HLT start with ```hlt```. I believe we have settled on ```hltSiStripClusters2ApproxClusters``` for that module.

File: Configuration/EventContent/python/EventContent_cff.py (hunk 4 above)
Comment (mandrenguyen) [reply Configuration/EventContent/python/EventContent_cff.py:643]: I just retested and it seems to be necessary as there is a drop `*_hlt*_*_*` command hidden somewhere in the appends.

**Comment by mandrenguyen** [comment]
+reconstruction

File: RecoLocalTracker/SiStripClusterizer/python/SiStripClusters2ApproxClusters_cff.py (hunk 2 above)
Comment (mandrenguyen) [reply RecoLocalTracker/SiStripClusterizer/python/SiStripClusters2ApproxClusters_cff.py:5]: Ok, so I keep this PR as it is. Thanks.

**Comment by Martin-Grunewald** [comment]
Concerning hltSiStripClusters2ApproxClusters, this will be the module label at HLT.
I assume your code is to use that module label as InputTag, correct?
(You should not run a hltSiStripClusters2ApproxClusters labelled producer outside HLT...., taht would break running HLT in the same step as your code)

**Comment by mandrenguyen** [comment]
> Concerning hltSiStripClusters2ApproxClusters, this will be the module label at HLT. I assume your code is to use that module label as InputTag, correct? (You should not run a hltSiStripClusters2ApproxClusters labelled producer outside HLT...., that would break running HLT in the same step as your code)

Yes, we need to specify it as an InputTag
We have also implemented a separate repacking step for testing, which mimics the job of the HLT. 
It will never be run with the HLT step.  To see how it looks:  runTheMatrix.py -l 140.6 -n -e

File: Configuration/EventContent/python/EventContent_cff.py [hunk 6]
```diff
@@ -634,6 +638,10 @@ def SwapKeepAndDrop(l):
 FEVTDEBUGHLTEventContent.outputCommands.append('keep *_*_MergedTrackTruth_*')
 FEVTDEBUGHLTEventContent.outputCommands.append('keep *_*_StripDigiSimLink_*')
 FEVTDEBUGHLTEventContent.outputCommands.append('keep *_*_PixelDigiSimLink_*')
+approxSiStripClusters.toModify(FEVTDEBUGHLTEventContent,
```
Comment (srimanob) [inline Configuration/EventContent/python/EventContent_cff.py:641]: Do we really need this line?
`FEVTDEBUGHLT` already calls `FEVTDEBUG` with
`FEVTDEBUGHLTEventContent.outputCommands.extend(FEVTDEBUGEventContent.outputCommands)`

https://github.com/cms-sw/cmssw/blob/master/Configuration/EventContent/python/EventContent_cff.py#L632

Comment (mandrenguyen) [reply Configuration/EventContent/python/EventContent_cff.py:641]: Yes, it's needed b/c there is a drop *_hlt*_*_* hidden in one of the append statements.

File: Configuration/PyReleaseValidation/python/relval_steps.py [hunk 7]
```diff
@@ -2111,6 +2115,28 @@ def lhegensim2018ml(fragment,howMuch):
...
+                        '--process':'REHLT'
+}
+
+steps['RECOHID22APPROXCLUSTERS']=merge([{ '--scenario':'pp',
+                                          '--conditions':'auto:run3_data_prompt',
+                                          '-s':'RAW2DIGI,L1Reco,RECO,DQM:@commonFakeHLT+@standardDQMFakeHLT',
+                                          '--datatier':'AOD,DQMIO',
+                                          '--eventcontent':'AOD,DQM',
+                                          '--era':'Run3_pp_on_PbPb_approxSiStripClusters',
+                                          '--repacked':'',
```
[the remark this answers is not in the document]
Comment (mandrenguyen) [reply Configuration/PyReleaseValidation/python/relval_steps.py:2135]: The `repacked` flag in `RECOHID22APPROXCLUSTERS` is actually redundant.   It's already present in `RECOHID15`, which is merged in both `RECOHID18APPROXCLUSTERS` and `RECOHID22APPROXCLUSTERS` .

**Comment by srimanob** [comment]
+Upgrade

**Comment by kskovpen** [comment]
+pdmv

**Comment by perrotta** [comment]
+1
- @cms-sw/daq-l2 has [signed](https://github.com/cms-sw/cmssw/pull/39883#issuecomment-1297095942) the identical backport PR in 12_5_X and not this one in the master: let assume that this master PR was just forgotten and bypass DAQ signature here. Please @smorovic complain here if there was any reason to sign the backport only (but please specify it in the comments next time!)

**Comment by perrotta** [comment]
merge

**Comment by smorovic** [comment]
+1
Yes, it was just ommitted by mistake. Thanks for pointing it out.

File: Configuration/StandardSequences/python/DigiToRaw_Repack_cff.py (hunk 1 above)
Comment (mmusich) [inline Configuration/StandardSequences/python/DigiToRaw_Repack_cff.py:80]: @mandrenguyen @icali 

This list is missing FED 1022, which is very important to make DCS checks in the many places (Tracker DQM and ALCA).

https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/DataFormats/OnlineMetaData/interface/DCSRecord.h#L4-L9

Why was it not included?

FYI:

@francescobrivio @silviodonato @missirol  (as far as I understand this will need to be fixed also at HLT)

Comment (francescobrivio) [reply Configuration/StandardSequences/python/DigiToRaw_Repack_cff.py:80]: Addessed for online in https://its.cern.ch/jira/browse/CMSHLT-2560 by @missirol.
I believe @mmusich will provide a PR to fix the list in the release as well (Marco let me know if you want me to do it).
